packages/stokhos/src/kokkos/Cuda/Stokhos_Cuda_CrsMatrix.hpp

```diff
@@ -61,7 +61,21 @@
 // Use new cuSPARSE SpMv/SpMM functions only in CUDA 11 or greater.
 // (while they exist in some versions of CUDA 10, they appear to not always
 // product correct results).
-#define USE_NEW_SPMV (CUSPARSE_VERSION >= 11000)
+#if CUSPARSE_VERSION >= 11201
+  #define USE_NEW_SPMV 1
+  #define NEW_SPMV_ALG_DEFAULTS \
+    cusparseSpMVAlg_t alg_spmv = CUSPARSE_SPMV_ALG_DEFAULT;
+  #define NEW_SPMM_ALG_DEFAULTS \
+    cusparseSpMMAlg_t alg_spmm = CUSPARSE_SPMM_ALG_DEFAULT;
+#elif CUSPARSE_VERSION >= 11000
+  #define USE_NEW_SPMV 1
+  #define NEW_SPMV_ALG_DEFAULTS \
+    cusparseSpMVAlg_t alg_spmv = CUSPARSE_MV_ALG_DEFAULT;
+  #define NEW_SPMM_ALG_DEFAULTS \
+    cusparseSpMMAlg_t alg_spmm = CUSPARSE_MM_ALG_DEFAULT;
+#else
+  #define USE_NEW_SPMV 0
+#endif
```

Comment:
Generally I am fine with this.  However, the logic here appears to be copied from KokkosKernels and my experience has been this logic isn't completely correct.  There is an important customer who builds with version 11.2.0 of the CUDA toolkit where the cuSPARSE version reports to be 11.3.1:
```
#define CUSPARSE_VER_MAJOR 11
#define CUSPARSE_VER_MINOR 3
#define CUSPARSE_VER_PATCH 1
#define CUSPARSE_VER_BUILD 68
#define CUSPARSE_VERSION (CUSPARSE_VER_MAJOR * 1000 + \
                          CUSPARSE_VER_MINOR *  100 + \
                          CUSPARSE_VER_PATCH)
```
which would choose the first branch, but uses `CUSPARSE_MV_ALG_DEFAULT` as in the second branch:
```
typedef enum {
    CUSPARSE_MV_ALG_DEFAULT = 0,
    CUSPARSE_COOMV_ALG      = 1,
    CUSPARSE_CSRMV_ALG1     = 2,
    CUSPARSE_CSRMV_ALG2     = 3
} cusparseSpMVAlg_t;
```
I've run into this issue several times with KokkosKernels, but have been too lazy to file a bug with them.  It would be annoying to have to fix this in two places though.

Reply:
Hi @etphipp. OK. I've done a search in Trilinos:

- In `Kokkos-kernels/perf_test/sparse/KokkosSparse_spmv_struct_tuning.cpp`, the conditional is `#if CUSPARSE_VERSION >= 11201`.
- In `Kokkos-kernels/sparse/tpls/KokkosSparse_spmv_tpl_spec_decl.hpp`, the conditional is `#if CUSPARSE_VERSION >= 11301`.
- In `Kokkos-kernels/sparse/tpls/KokkosSparse_spmv_mv_tpl_spec_decl.hpp`, the conditional is `#if CUSPARSE_VERSION < 11010`.
- In `mule/test/scaling/MatvecKernelDriver.cpp`, the conditional is `#if CUSPARSE_VERSION >= 11201`.

So you think we should find out exactly which version changed it in which case MM/MV and then do two simultaneous PRs, one in Trilinos (for Stokhos and Muelu) and one in Kokkos-kernels?

Reply:
Ideally, yes, I guess.